### PR TITLE
2024 roadmap post

### PR DESCRIPTION
**Pull request by jmcook1186**
Adds post detailing the 2024 roadmap
Requires some input from Portal Network team members:

- general review of whole doc
- client specific information in the `2024 Workstreams` section
- any other information you want to add

- [x] Trin details
- [x] Fluffy details
- [x] Ultralight details
- [x] Glados details
- [x] General review

### CODE REVIEW

File: pages/posts/roadmap-2024.md [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+---
+
+# Portal Network Roadmap Draft 2024
+
+This document aims to outline the 2024 core team & product roadmap for Portal Network.
+
+
+## Team description
+
+The Portal Network core team is a small (<15 person) collaborative effort focused on bulding a decentralized peer-to-peer network for serving various types of critical Ethereum data. The Portal project is composed of three independent client teams each implementing the Portal Network specification in different languages: Trin (Rust), Ultralight (Typescript) and Fluffy (Nim), plus cross-client efforts to build out tooling and testing infrastructure.
```
Comment (pipermerriam) [inline pages/posts/roadmap-2024.md:17]:
```suggestion
The Portal Network core team is a small (<15 person) collaborative effort focused on bulding a decentralized peer-to-peer network for storing and serving data from the execution and consensus layers of the Ethereum protocol. The Portal project is composed of three independent client teams each implementing the Portal Network specification in different languages: Trin (Rust), Ultralight (Typescript) and Fluffy (Nim), plus cross-client efforts to build out tooling and testing infrastructure. 
```

File: pages/posts/roadmap-2024.md [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+
+These sub-protocols are prioritized because together they provide verifiable access to the full Ethereum dataset in a manner that is fully decentralized and accessible on resource constrained devices. 
+
+Specifically: 
+
+- The History sub-protocol enables Ethereum execution clients to expire historical data via EIP-4444, safe in the knowledge that the data is available on request from the Portal Network. 
+- The State sub-protocol gives light clients access to Ethereum stata data, exposing methods such as `eth_getProof` that eliminate the need for trusted third party RPC providers. 
+- The Beacon sub-protocol gives light clients access to the latest block headers so that they can be confident they are following the canonical Ethereum chain.
+
+Shipping these features will enable ***major upgrades for the Ethereum ecosystem as a whole***. These featues have been in development for several years and have been through several rounds of refinement and course correction. Now Portal Network is on the brink of shipping these sub-networks into production.
```
Comment (pipermerriam) [inline pages/posts/roadmap-2024.md:43]:
```suggestion
Shipping these features will enable ***major upgrades for the Ethereum ecosystem as a whole***. These features have been in development for several years and have been through several rounds of refinement and course correction. Now Portal Network is on the brink of shipping these sub-networks into production. 
```

File: pages/posts/roadmap-2024.md [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+- **Objective 4: Portal developers ship major new features that change how normal users interact with Ethereum**
+
+    - Key result 1: Portal team have beta implementations of canonical transaction index network
+    
+    - Key result 2: Portal team have agreed and begun to implement a design for a two-way bridge to Ethereum to enable transaction forwarding from Poral --> Ethereum
+
+
+## Where are we now?
+
+The Portal Network has existed for 5 years, during which time there have been substantial course corrections and redesigns in response tothe evolving wider Ethereum landscape. Today, Portal Network is not widely used, but it is on the precipice of launching highly impactful infrastructure that will benefit large parts of the Ethereum ecosystem. 2024 is expected to be a pivotal year, where we focus on moving what were previously R&D projects into production and focusing on growing adoption for our core features: the History network, State network and Beacon network.
```
Comment (pipermerriam) [inline pages/posts/roadmap-2024.md:123]:
```suggestion
The Portal Network has existed for 5 years, during which time there have been substantial course corrections and redesigns in response to the evolving wider Ethereum landscape. Today, Portal Network actively deploying infrastructure that will benefit large parts of the Ethereum ecosystem. 2024 is expected to be a pivotal year, where we focus on moving what were previously R&D projects into production and focusing on growing adoption for our core features: the History network, State network and Beacon network.
```

File: pages/posts/roadmap-2024.md [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+As well as successes, there were some frustrations for the Portal Network team in 2023:
+
+
+## 2024 workstreams
+
+To achieve our mission, we’re focused on 5 major workstreams that comprise our roadmap.
+
+### 1) Trin
+
+- The Trin team comprises N developers.
```
Comment (pipermerriam) [inline pages/posts/roadmap-2024.md:164]:
```suggestion
- The Trin team is comprised of 5 developers.
```

File: pages/posts/roadmap-2024.md [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+- Trin developers also work on tooling such as Glados and maintaining the testing infrastructure.
```
Comment (pipermerriam) [inline pages/posts/roadmap-2024.md:165]:
```suggestion
- Trin developers also work on tooling such as Glados and maintaining the hive testing infrastructure.
```

File: pages/posts/roadmap-2024.md (hunk 1 above)
Comment (njgheorghita) [inline pages/posts/roadmap-2024.md:17]: `building`

File: pages/posts/roadmap-2024.md [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+        - Portal Hive has 100% test coverage for all core functionality for all three networks across all three flagship Portal clients
+
+
+- **Objective 5: Communicate effectively with users and developers**
+    - Key result 1: 
+        - Github issues and comments across the Portal specs, Glados, and client repositories are always responded to (even if just an acknowledgement of receipt) in less than 3 days.
+    - Key result 2: 
+        - New features and milestones are reported via the Portal blog within 1 month of final PR meging merged.
+    - Key result 3:
+        - At least 5 presentations focused on portal Network are delivered at major Ethereum events in 2024, including Devcon.
```
Comment (njgheorghita) [inline pages/posts/roadmap-2024.md:97]: `Portal Network`

File: pages/posts/roadmap-2024.md [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+
+### 2025 and beyond
+
+These are longer term objectives that might be researched and discussed during 2024 but are not expected to be shipped until at least 2025. These will be revisited during 2025 roadmap planning.
+
+- **Objective 1: Portal Network is widely accepted as the default source for light client execution and consensus data**
+    - \> 10 apps/light clients have replaced third party centralized RPC providers with Portal Network.
+
+- **Objective 2: Portal Network is widely considered critical Ethereum infrastructure** 
+    - Key result 1: Portal mode is available is most Ethereum execution clients, either enablign them to aggressively prune their local data or to accelerate their sync times.
```
Comment (njgheorghita) [inline pages/posts/roadmap-2024.md:108]: `enabling`

File: pages/posts/roadmap-2024.md [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+
+- **Objective 3: Portal Network is supporting major new developments for Ethereum**
+
+    - Key result 1: Portal Network serves data enabling e.g. stateless clients, SNARKified EVM, or other substantial Ethereum upgrades from the protocol roadmap. 
+
+- **Objective 4: Portal developers ship major new features that change how normal users interact with Ethereum**
+
+    - Key result 1: Portal team have beta implementations of canonical transaction index network
+    
+    - Key result 2: Portal team have agreed and begun to implement a design for a two-way bridge to Ethereum to enable transaction forwarding from Poral --> Ethereum
```
Comment (njgheorghita) [inline pages/posts/roadmap-2024.md:118]: `Portal --> `

File: pages/posts/roadmap-2024.md [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+
+- **Objective 1: Portal Network is widely accepted as the default source for light client execution and consensus data**
+    - \> 10 apps/light clients have replaced third party centralized RPC providers with Portal Network.
+
+- **Objective 2: Portal Network is widely considered critical Ethereum infrastructure** 
+    - Key result 1: Portal mode is available is most Ethereum execution clients, either enablign them to aggressively prune their local data or to accelerate their sync times.
+
+- **Objective 3: Portal Network is supporting major new developments for Ethereum**
+
+    - Key result 1: Portal Network serves data enabling e.g. stateless clients, SNARKified EVM, or other substantial Ethereum upgrades from the protocol roadmap.
```
Comment (njgheorghita) [inline pages/posts/roadmap-2024.md:112]: I think you could add support for verkle trees here

File: pages/posts/roadmap-2024.md [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+
+
+## Where are we now?
+
+The Portal Network has existed for 5 years, during which time there have been substantial course corrections and redesigns in response tothe evolving wider Ethereum landscape. Today, Portal Network is not widely used, but it is on the precipice of launching highly impactful infrastructure that will benefit large parts of the Ethereum ecosystem. 2024 is expected to be a pivotal year, where we focus on moving what were previously R&D projects into production and focusing on growing adoption for our core features: the History network, State network and Beacon network.
+
+In the first instance, users will be expected to run Portal clients to expose a subset of Ethereum RPC methods. The list of methods will grow steadily over the first half of 2024 until almost all of the `eth` namespace methods can be served from Portal data. however, we also intend to collaborate with Ethereum client teams to help them to implement the Portal specification directly in existing execution clients, circumventing the need to run additional clients.
```
Comment (njgheorghita) [inline pages/posts/roadmap-2024.md:125]: `data. However, we ...`

File: pages/posts/roadmap-2024.md (hunk 4 above)
Comment (njgheorghita) [inline pages/posts/roadmap-2024.md:164]:
```
- The Trin team comprises 5 developers.
- Trin developers also maintain tooling such as Glados, Portal-Hive, and Portal-Bridge.
- The Trin team is especially focused on delivering a stable 4444s History network with validation for post-merge data via the Beacon network.
- Specific challenges they might encounter in delivering the 2024 roadmap include...
  - Stabilizing `utp` transport tooling / debugging async network issues
  - Stabilizing the History/Beacon networks, while supporting the alpha State network
  - Scaling pains of monitoring tools
  - Network failures due to unmitigated attack vectors
```

File: pages/posts/roadmap-2024.md [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+
+### 2024 
+
+- **Objective 1: Ship a viable solution for historical data storage**
+    - Key result 1:
+        -  Portal History subprotocol is fully featured according to the canonical specification and there is feature parity across all three Portal clients. 
+    - Key result 2:
+        - \>=1 execution client is using Portal History network to support EIP-4444
+    - Key result 3:
+        - History network never drops below 99% score on GladOS
```
Comment (kdeme) [inline pages/posts/roadmap-2024.md:60]:
```suggestion
        - History network never drops below 99% score on Glados
```

To be consistent.

(Or change it to GLaDOS everywhere ;-) )

File: pages/posts/roadmap-2024.md [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+        - \>=1 application or light client using Portal Network to follow the head of the Beacon chain
+    - Key result 4:
+        - \>=1 execution client using Portal Beacon network to accelerate their syncing
+
+
+- **Objective 3: Ship a viable solution for decentralized access to state data**:
+    - Key result 1:
+        -  Portal State subprotocol is fully featured according to the canonical specification and there is feature parity across all three Portal clients. 
+    -  Key result 2:
+        -  State network never drops below 99% score on Glados
```
Comment (kdeme) [inline pages/posts/roadmap-2024.md:80]: For recent states only perhaps...?

File: pages/posts/roadmap-2024.md [hunk 13]
```diff
@@ -0,0 +1,219 @@
...
+Shipping these features will enable ***major upgrades for the Ethereum ecosystem as a whole***. These featues have been in development for several years and have been through several rounds of refinement and course correction. Now Portal Network is on the brink of shipping these sub-networks into production. 
+
+This is why ***2024 is a set to be a standout year for Portal Network.***
+
+Longer term, there are proposals for more sub-protocols that serve granular transaction data, blob data, proofs etc, but these are out of scope for 2024.
+
+
+## High level goals
+
+### 2024
```
Comment (kdeme) [inline pages/posts/roadmap-2024.md:52]: I understand that these goals are goals and not commitments.

But are the key results to be interpreted from good to great based on their numbering?

File: pages/posts/roadmap-2024.md [hunk 14]
```diff
@@ -0,0 +1,219 @@
...
+        - Glados maintains 99% uptime and is able to handle high volume of traffic (up to 10^5 visits/day)
+    - Key result 2:
+        - Portal Hive has 100% test coverage for all core functionality for all three networks across all three flagship Portal clients
+
+
+- **Objective 5: Communicate effectively with users and developers**
+    - Key result 1: 
+        - Github issues and comments across the Portal specs, Glados, and client repositories are always responded to (even if just an acknowledgement of receipt) in less than 3 days.
+    - Key result 2: 
+        - New features and milestones are reported via the Portal blog within 1 month of final PR meging merged.
```
Comment (kdeme) [inline pages/posts/roadmap-2024.md:95]:
```suggestion
        - New features and milestones are reported via the Portal blog within 1 month of final PR being merged.
```

File: pages/posts/roadmap-2024.md (hunk 3 above)
Comment (kdeme) [inline pages/posts/roadmap-2024.md]: Is it not rather about 3 years? And before focus was rather statelessness and beam sync?

File: pages/posts/roadmap-2024.md (hunk 6 above)
Comment (jmcook1186) [inline pages/posts/roadmap-2024.md:97]:
```suggestion
        - At least 5 presentations focused on Portal Network are delivered at major Ethereum events in 2024, including Devcon.
```

File: pages/posts/roadmap-2024.md (hunk 7 above)
Comment (jmcook1186) [reply pages/posts/roadmap-2024.md:108]:
```suggestion
    - Key result 1: Portal mode is available is most Ethereum execution clients, either enabling them to aggressively prune their local data or to accelerate their sync times.
```

File: pages/posts/roadmap-2024.md (hunk 9 above)
Comment (jmcook1186) [reply pages/posts/roadmap-2024.md:112]:
```suggestion
    - Key result 1: Portal Network serves data enabling e.g. stateless clients, SNARKified EVM, Verkle trees, or other substantial Ethereum upgrades from the protocol roadmap. 
```

File: pages/posts/roadmap-2024.md (hunk 8 above)
Comment (jmcook1186) [reply pages/posts/roadmap-2024.md:118]:
```suggestion
    - Key result 2: Portal team have agreed and begun to implement a design for a two-way bridge to Ethereum to enable transaction forwarding from Portal --> Ethereum
```

File: pages/posts/roadmap-2024.md (hunk 10 above)
Comment (jmcook1186) [reply pages/posts/roadmap-2024.md:125]:
```suggestion
In the first instance, users will be expected to run Portal clients to expose a subset of Ethereum RPC methods. The list of methods will grow steadily over the first half of 2024 until almost all of the `eth` namespace methods can be served from Portal data. However, we also intend to collaborate with Ethereum client teams to help them to implement the Portal specification directly in existing execution clients, circumventing the need to run additional clients. 
```

File: pages/posts/roadmap-2024.md [hunk 15]
```diff
@@ -0,0 +1,223 @@
...
+  - Stabilizing the History/Beacon networks, while supporting the alpha State network
+  - Scaling pains of monitoring tools
+  - Network failures due to unmitigated attack vectors
+ 
+### 2) Fluffy
+
+- The Fluffy team comprises N developers.
+- Fluffy developers also work on tooling such as Glados and maintaining the testing infrastructure.
+- The Fluffy team is especially focused on....
+- Specific challenges they might encounter in delivering the 2024 roadmap include...
```
[the remark this answers is not in the document]
Comment (kdeme) [reply pages/posts/roadmap-2024.md:178]: For glados and portal-hive I specifically changed it to `assist` as I find work or help to strong.
We are far from the work that Trin developers do there, so I try to tone it down a bit :)

File: pages/posts/roadmap-2024.md (hunk 13 above)
Comment (jmcook1186) [reply pages/posts/roadmap-2024.md:52]: The order is arbitrary - there is no prioritisation to be inferred from the numbering.

File: pages/posts/roadmap-2024.md (hunk 12 above)
Comment (jmcook1186) [reply pages/posts/roadmap-2024.md:80]: These are suggested OKRs - I'd be delighted if you tweak them according to what you think is reasonable.

File: pages/posts/roadmap-2024.md (hunk 3 above)
Comment (jmcook1186) [reply pages/posts/roadmap-2024.md]: I am not sure - I guess the start is fuzzy as it grew from other earlier projects. @pipermerriam ?

File: pages/posts/roadmap-2024.md [hunk 16]
```diff
@@ -0,0 +1,228 @@
...
+This is why ***2024 is a set to be a standout year for Portal Network.***
+
+Longer term, there are proposals for more sub-protocols that serve granular transaction data, blob data, proofs etc, but these are out of scope for 2024.
+
+
+## High level goals
+
+### 2024 
+
+- **Objective 1: Ship a viable solution for historical data storage**
```
Comment (mrferris) [inline pages/posts/roadmap-2024.md:54]:
```suggestion
- **Objective 1: Ship a viable solution for decentralized access to historical data**
```
The different wording between 1 & 2/3 might have been implying that historical data access doesn't need to be decentralized or accessible

File: pages/posts/roadmap-2024.md [hunk 17]
```diff
@@ -0,0 +1,228 @@
...
+    - Key result 1:
+        -  Portal State subprotocol is fully featured according to the canonical specification and there is feature parity across all three Portal clients. 
+    -  Key result 2:
+        -  State network never drops below 99% score on Glados
+    -  Key result 3:
+        - \>=1 application or client using Portal State network to request `eth_getProof` data
+
+- **Objective 4: Develop and maintain reliable tooling**
+    - Key result 1: 
+        - Glados maintains 99% uptime and is able to handle high volume of traffic (up to 10^5 visits/day)
```
Comment (mrferris) [inline pages/posts/roadmap-2024.md:86]: The above objectives already imply it, but maybe we want to explicitly add: 
`- Glados provides network audit statistics for the history, beacon, and state networks`

File: pages/posts/roadmap-2024.md [hunk 18]
```diff
@@ -0,0 +1,228 @@
...
+- **Objective 4: Portal developers ship major new features that change how normal users interact with Ethereum**
+
+    - Key result 1: Portal team have beta implementations of canonical transaction index network
+    
+    - Key result 2: Portal team have agreed and begun to implement a design for a two-way bridge to Ethereum to enable transaction forwarding from Portal --> Ethereum
+
+
+## Where are we now?
+
+The Portal Network has existed for 5 years, during which time there have been substantial course corrections and redesigns in response to the evolving wider Ethereum landscape. Today, Portal Network actively deploying infrastructure that will benefit large parts of the Ethereum ecosystem. 2024 is expected to be a pivotal year, where we focus on moving what were previously R&D projects into production and focusing on growing adoption for our core features: the History network, State network and Beacon network.
```
Comment (mrferris) [inline pages/posts/roadmap-2024.md:123]:
```suggestion
The Portal Network has existed for 5 years, during which time there have been substantial course corrections and redesigns in response to the evolving wider Ethereum landscape. Today, Portal Network is actively deploying infrastructure that will benefit large parts of the Ethereum ecosystem. 2024 is expected to be a pivotal year, where we focus on moving what were previously R&D projects into production and focusing on growing adoption for our core features: the History network, State network and Beacon network.
```

File: pages/posts/roadmap-2024.md [hunk 19]
```diff
@@ -0,0 +1,228 @@
...
+
+- The Ultralight team comprises N developers.
+- Ultralight developers also work on tooling such as Glados and maintaining the testing infrastructure.
+- The Ultralight team is especially focused on....
+- Specific challenges they might encounter in delivering the 2024 roadmap include...
+...
+### 4) Glados
+
+- Glados is worked on by developers from the three main client teams, with some individuals taking more prominent roles than others. 
+- Scaling is going to be a challenge for Glados in 2024. The tool may have to migrate to an EF managed server so we can benefit from devops support and handle large volumes of traffic.
```
Comment (mrferris) [inline pages/posts/roadmap-2024.md:195]: Glados currently runs on an EF managed server, specifically managed by Paul on the devops team. Not sure if this is referring to something else. 
```suggestion
- Scaling Glados from one to three Portal networks while simultaneously scaling up the expected traffic volume will be a challenge in 2024. The primary bottleneck in this scaling will be the Postgres DB that holds all of Glados's data. The primary path to accomplishing this scaling will be to continue improving Glados's DB queries and indexing. 
```

File: pages/posts/roadmap-2024.md [hunk 20]
```diff
@@ -0,0 +1,228 @@
...
+  - Balancing time and efforts to further develop the Fluffy version of a Portal-bridge and actively run it on the network.
+  - Client/network robustness, resilience and performance as the networks scale up in terms of amounts of nodes and data + the capability to properly monitor this. 
+  
+
+### 3) Ultralight
+
+- The Ultralight team comprises N developers.
+- Ultralight developers also work on tooling such as Glados and maintaining the testing infrastructure.
+- The Ultralight team is especially focused on....
+- Specific challenges they might encounter in delivering the 2024 roadmap include...
```
Comment (acolytec3) [inline pages/posts/roadmap-2024.md:190]:
```suggestion
- The Ultralight team comprises 2 developers.
- Ultralight developers also work on tooling such as Portal-Hive and maintaining the testing infrastructure.
- The Ultralight team is especially focused on expanding the current proof-of-concept State Network implementation into a fully-featured spec complete version
- Specific challenges they might encounter in delivering the 2024 roadmap include...
  - Manpower - the current team members are both part of the EthereumJS team and work part time on other team responsibilities
  - Brower network incompatibilities - the current proof of concept implementation of the browser client is dependent on proxy services to connect to the rest of the network (due to discv5 being UDP exclusive) so finding a browser-friendly transport layer would be ideal to make browser light clients a reality
```

File: pages/posts/roadmap-2024.md (hunk 12 above)
Comment (kdeme) [reply pages/posts/roadmap-2024.md:80]: Well, I guess it is not yet defined what Glados score for state network exactly means.

If it means all state history data audited, that for sure seems though. But ok, in its current form, that is not stated, nor is it not not stated. So we could keep it intentionally vague :)

But perhaps: "State network audits on Glados for recent states never drops below 99%."

File: pages/posts/roadmap-2024.md [hunk 21]
```diff
@@ -0,0 +1,234 @@
+
+---
+title: Portal Network 2024 roadmap
+date: 2024/2/1
+description: What does 2024 have in store for Portal Network?
+tag: project
+author: @jmcook1186
```
Comment (jmcook1186) [inline pages/posts/roadmap-2024.md:7]:
```suggestion
author: jmc
```

**Pull request merged**